---
layout: post
title: Big O and Algorithmic Efficiency
description: Big O and Algorithmic Efficiency HW and Popcorn Hacks
type: issues
comments: true
---

### Popcorn Hack 2

In [1]:
import time
import random

# Generate a large sorted list
data_size = 10000000
sorted_data = sorted(random.sample(range(100000000), data_size))

# Target to find (worst case for linear search)
target = sorted_data[-1]  # Last element

# O(n) - Linear Search
def linear_search(arr, target):
    for i, element in enumerate(arr):
        if element == target:
            return i
    return -1

# O(log n) - Binary Search
def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    
    return -1

# Compare performance
print("Testing with data size:", data_size)

start = time.time()
linear_result = linear_search(sorted_data, target)
linear_time = time.time() - start
print(f"Linear search: {linear_time:.6f} seconds")

start = time.time()
binary_result = binary_search(sorted_data, target)
binary_time = time.time() - start
print(f"Binary search: {binary_time:.6f} seconds")

print(f"Binary search is approximately {linear_time/binary_time:.0f}x faster")

Testing with data size: 10000000
Linear search: 3.914667 seconds
Binary search: 0.000063 seconds
Binary search is approximately 62194x faster


### 1) What is the time complexity of each algorithm?
- #### Linear search has a time complexity of O(n) because it checks each element one by one until it finds the target
- #### Binary search has a time complexity of O(log n) because it divides the search space in half with each step

### 2) How many times faster is binary search than linear search?
- #### Binary search is approximately 62,194 times faster than linear search in this test with 10 million elements

### 3) What happens if you increase data_size to 20000000?
- #### Linear search will take about twice as long, because it scales linearly with the size of the list
- #### Binary search will only take a slightly longer time, since log base 2 of 20 million is only a bit more than of 10 million